src/test/app/DepositAuth_test.cpp

```diff
@@ -724,14 +792,506 @@ struct DepositPreauth_test : public beast::unit_test::suite
         }
     }
 
+    void
+    testCredentials()
+    {
+        using namespace jtx;
+
+        const char credType[] = "abcde";
+        Account const issuer{"issuer"};
+        Account const alice{"alice"};
+        Account const bob{"bob"};
+        Account const john("john");
+        Account const maria{"maria"};
+
+        {
+            testcase("Payment with credentials.");
+
+            Env env(*this);
+
+            env.fund(XRP(5000), issuer, alice, bob);
+            env.close();
+
+            // Issuer create credentials, but Alice didn't accept them yet
+            env(credentials::create(alice, issuer, credType));
+            env.close();
+
+            // Get the index of the credentials
+            auto const jv =
+                credentials::ledgerEntry(env, alice, issuer, credType);
+            std::string const credIdx = jv[jss::result][jss::index].asString();
+
+            // Bob require preauthorization
+            env(fset(bob, asfDepositAuth));
+            env.close();
+
+            // Bob will accept payements from accounts with credentials signed
+            // by 'issuer'
+            env(deposit::authCredentials(bob, {{issuer, credType}}));
+            env.close();
+
+            auto const jDP =
+                ledgerEntryDepositPreauth(env, bob, {{issuer, credType}});
+            BEAST_EXPECT(
+                jDP.isObject() && jDP.isMember(jss::result) &&
+                !jDP[jss::result].isMember(jss::error) &&
+                jDP[jss::result].isMember(jss::node) &&
+                jDP[jss::result][jss::node].isMember("LedgerEntryType") &&
+                jDP[jss::result][jss::node]["LedgerEntryType"] ==
+                    jss::DepositPreauth);
+
+            // Alice can't pay - empty credentials array
+            {
+                auto jv = pay(alice, bob, XRP(100));
+                jv[sfCredentialIDs.jsonName] = Json::arrayValue;
+                env(jv, ter(temMALFORMED));
+                env.close();
+            }
+
+            // Alice can't pay - not accepeted credentials
+            env(pay(alice, bob, XRP(100)),
+                credentials::ids({credIdx}),
+                ter(tecBAD_CREDENTIALS));
+            env.close();
+
+            // Alice accept the credentials
+            env(credentials::accept(alice, issuer, credType));
+            env.close();
+
+            // Now Alice can pay
+            env(pay(alice, bob, XRP(100)), credentials::ids({credIdx}));
+            env.close();
+
+            // Alice can pay Maria without depositPreauth enabled
+            env(pay(alice, maria, XRP(250)), credentials::ids({credIdx}));
+            env.close();
```

Comment:
Some additional tests to include:
* Alice doesn't provide the (valid) credentials
* Credential doesn't belong to tx sender

Reply:
present in testCredentials()